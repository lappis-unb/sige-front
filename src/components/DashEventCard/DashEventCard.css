#DashEventCard p {
  font-size: 20px;
}

.card-title {
  margin-left: 8px;
}

.card-overflow {
  overflow-y: auto;
  overflow-x: hidden;
}

.card-inactive {
  background-color: #00080f;
  border: 2px solid #001324;
  color: rgba(255, 255, 255, 0.2);
}

.card-active-critical {
  background-color: #3d0202;
  color: white;
}

.card-active-warning {
  background-color: #4d3e00;
  color: white;
}

.card-title {
  font-size: 30px;
  line-height: 40px;
  letter-spacing: 0.24px;
}
