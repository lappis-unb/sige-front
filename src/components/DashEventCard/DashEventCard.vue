<template>
  <div class="base" id="DashEventCard">
    <q-card flat class="evt-card-size card-overflow" :class="IS_ACTIVE ? cardTypeStyle : 'card-inactive'">
      <q-card-section class="text-center card-title">
        <span v-if="IS_ACTIVE">
          <q-icon :name="`${ICON_PATH}${icon}.svg`" />
        </span>
        <strong class="card-title">{{ eventTile }}</strong>
      </q-card-section>
      <q-card-section class="text-center q-px-xs" v-if="IS_ACTIVE">
        <p v-for="event in campusEvents.slice(0, MAX_LOCATION_POSSIBLE)" :key="event.id" class="q-my-xs">
          {{ getCampusInformation(event) }}
        </p>
        <p v-if="campusEvents.length > MAX_LOCATION_POSSIBLE" class="bottom q-mt-sm">
          <strong> E mais {{ campusEvents.length - MAX_LOCATION_POSSIBLE }} outro(s) </strong>
        </p>
      </q-card-section>
      <q-card-section v-else class="absolute-center">
        <strong>Nenhum ponto</strong>
      </q-card-section>
    </q-card>
  </div>
</template>

<script>
export default {
  name: 'DashEventCard',

  data() {
    return {
      IS_ACTIVE: this.campusEvents.length !== 0,
      MAX_LOCATION_POSSIBLE: 8,
      ICON_PATH: 'img:statics/ic_ocorrencia_',
      ACCEPTED_VOLTAGES: /^voltage_[abc]/,
      LOCATION_NAME_MAX_CHAR: 15
    }
  },

  props: {
    eventTile: String,
    cardTypeStyle: String,
    campusEvents: Array,
    icon: Number
  },

  methods: {
    getCampusInformation(campusInformation) {
      const { data: voltages, location } = campusInformation

      if (location === '') return 'Sem localização';

      let locationAndCampus = this.formatLocationName(campusInformation);
      if (this.name !== 'Falha de Comunicação') locationAndCampus += ` - ${this.concatVoltageTypes(voltages)}`;
      return locationAndCampus;
    },

    formatLocationName(campusInformation) {
      const { location, campus } = campusInformation

      const _location =
        location.length > this.LOCATION_NAME_MAX_CHAR
          ? location.slice(0, this.LOCATION_NAME_MAX_CHAR) + '...'
          : location

      return `${_location[0].toUpperCase() + _location.slice(1)} (${campus})`
    },

    concatVoltageTypes(voltages) {
      const voltageTypes = []

      Object.keys(voltages).forEach((voltageType) => {
        const lastChar = voltageType[voltageType.length - 1]
        if (voltageType.match(this.ACCEPTED_VOLTAGES)) voltageTypes.push(lastChar.toUpperCase())
      })

      return voltageTypes
    }
  }
}
</script>

<style>
@import './DashEventCard.css';
</style>
