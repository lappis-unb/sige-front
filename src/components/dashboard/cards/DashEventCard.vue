<template>
  <div class="base">
    <q-card
      flat
      class="evt-card-size"
      :class="isActive ? activeClass : 'card-inactive'">

      <q-card-section
        class="text-center card-title q-px-none">
        <span v-if="isActive">
          <q-icon :name="`img:statics/icons/ic_ocorrencia_${icon}.svg`"/>
        </span> {{name}}
      </q-card-section>

      <q-card-section
        class="text-center q-px-xs q-pt-none"
        v-if="isActive">

        <div v-if="eventsList.length < 9">
          <p v-for="ev in eventsList" :key="ev.id"
            class="q-my-xs">
            {{ formatText(ev) }}
          </p>
        </div>

        <div v-else>
          <p v-for="ev in eventsList.slice(0, 8)" :key="ev.id"
            class="q-my-none">
            {{ formatText(ev) }}
          </p>
          <p class="bottom q-mt-sm">
            E mais {{ eventsList.length - 8 }} outro(s)
          </p>
        </div>

      </q-card-section>

      <q-card-section
        v-else
        class="absolute-center">
        <h6 class="text-center">
          Nenhum ponto
        </h6>
      </q-card-section>

    </q-card>
  </div>
</template>

<script>
export default {
  name: 'DashEventCard',

  data () {
    return {
      isActive: this.eventsList.length !== 0
    }
  },

  props: {
    name: String,
    activeClass: String,
    eventsList: Array,
    icon: Number
  },

  methods: {
    formatText (eventObj) {
      if (eventObj.location === '') {
        return 'Sem localização'
      }

      const strBreakpoint = 100
      let splittedLocation = eventObj.location.length > strBreakpoint ? eventObj.location.slice(0, strBreakpoint) + '...' + eventObj.location.slice(-strBreakpoint) : eventObj.location
      let formattedStr = ''

      switch (this.name) {
        case 'Queda de Fase':
          formattedStr = `${splittedLocation[0].toUpperCase() + splittedLocation.slice(1)} (${eventObj.campus}) - ${this.formatDrops(eventObj)}`
          break

        case 'Falha de Comunicação':
          formattedStr = `${splittedLocation[0].toUpperCase() + splittedLocation.slice(1)} (${eventObj.campus.toUpperCase()})`
          break

        case 'Máximo Consumo':
          // Must be implemented when we can measure this
          break

        default:
          formattedStr = `${splittedLocation[0].toUpperCase() + splittedLocation.slice(1)} (${eventObj.campus}) - ${this.formatDrops(eventObj)}`

          break
      }

      return formattedStr
    },

    formatDrops (obj) {
      let drops = Object.keys(obj.data)
      let d = []
      drops.forEach(label => {
        switch (label) {
          case 'voltage_a':
            d.push('A')
            break
          case 'voltage_b':
            d.push('B')
            break
          case 'voltage_c':
            d.push('C')
            break
          default:
            d.push(label)
            break
        }
      })
      return d
    }
  }

}
</script>

<style lang="scss" scoped>
  .card-inactive {
    background-color: #00080f;
    border: 2px solid #001324;
    color: rgba(255, 255, 255, 0.2);
  }

  .card-active-critical {
    background-color: #3d0202;
    color: white;
  }

  .card-active-warning {
    background-color: #4d3e00;
    color: white;
  }

  p {
    // font-size: 22px;
    font-size: 20px;
  }

  .card-title {
    font-size: 30px;
    // font-size: 32px;
    line-height: 40px;
    letter-spacing: .24px;
  }

</style>
