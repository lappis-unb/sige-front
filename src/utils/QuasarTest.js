import Vuex from 'vuex'
import Vue from 'vue'
import { createLocalVue, shallowMount, mount } from '@vue/test-utils';
import Quasar from 'quasar';

class QuasarTest {
  
  _localVue;

  constructor(){
    this.setUpQuasar()
  }
  
  setUpQuasar() {
    this._localVue = createLocalVue()
    Vue.use(Quasar)
    this._localVue.use(Vuex)
    this._localVue.use(Quasar)
  }

  shallowMount (Component, propsData = {}) {
    const localVue = this._localVue;
    return shallowMount(Component, {
      localVue,
      propsData
    });
  }

  mount (Component, propsData = {}) {
    const localVue = this._localVue;
    return mount(Component, {
      localVue,
      propsData
    });
  }
}

export default new QuasarTest();