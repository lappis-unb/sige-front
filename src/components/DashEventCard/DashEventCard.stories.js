import DashEventCard from './index.js'

export default {
  title: 'Dash Event Card',
  component: DashEventCard,
  argTypes: {
    eventTile: { type: 'string' },
    cardTypeStyle: { type: 'string' },
    campusEvents: { type: 'array' },
    icon: { type: 'number' },
  }
}

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { DashEventCard },
  template: '<dash-event-card v-bind="$props"/>'
})

export const Default = Template.bind({})
Default.args = {
  eventTile: 'Queda de Fase',
  cardTypeStyle: 'card-active-warning',
  campusEvents: [
    {
      id: 1,
      location: 'Gama',
      campus: 'FGA',
      transductor: 'FGA - 2',
      data: {
        voltage_a: 1, 
        voltage_b: 1, 
        voltage_c: 1, 
      },
    },
  ],
  icon: 1
}