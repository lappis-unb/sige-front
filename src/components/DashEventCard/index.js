import DashEventCard from "./DashEventCard.vue";
export default DashEventCard;