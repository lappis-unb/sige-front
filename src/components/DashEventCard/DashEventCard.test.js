import QuasarTest from "src/utils/QuasarTest";
import DashEventCard from '.';

const PROPS_DATA = {
  eventTile: 'Queda de Fase Teste',
  cardTypeStyle: 'card-active',
  campusEvents: [
    {
      id: 1,
      location: 'Gama',
      campus: 'FGA',
      transductor: 'FGA - 2',
      data: {
        voltage_a: 1, 
        voltage_b: 1, 
        voltage_c: 1, 
      },
    },
  ],
  icon: 1
}

describe('DashEventCard', () => {
  describe("Snapshot", () => {
    describe.each([
      {cardTypeStyle: "card-active"},
      {cardTypeStyle: "card-inactive"},
      {cardTypeStyle: "card-active-critical"},
      {cardTypeStyle: "card-active-warning"},
    ])("cardTypeStyle", ({cardTypeStyle}) => {
      it(`Teste da quando prop de estilo é ${cardTypeStyle}`, () => {
        const wrapper = QuasarTest.shallowMount(DashEventCard, {...PROPS_DATA, cardTypeStyle});
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });
  describe("Props", ()=>{
    describe("campusEvents", ()=>{
      it("A localização deve ter seu nome cortado quando muito longo", () => {
        const campusEvents = [{location: "Cidade com nome muito longo", data: undefined, campus: undefined}]
        
        const wrapper = QuasarTest.shallowMount(DashEventCard, {...PROPS_DATA, campusEvents});
        const cityName = wrapper.find('#DashEventCardCampus')

        expect(cityName.text()).toBe("Cidade com nome...");
      })

      it("A localização ser concatenda com o campus e as voltagens", () => {
        const wrapper = QuasarTest.shallowMount(DashEventCard, PROPS_DATA);
        const cityName = wrapper.find('#DashEventCardCampus')
  
        expect(cityName.text()).toBe("Gama (FGA) - A,B,C");
      })

      it("Caso a localização seja indefinida deve retornar 'Sem localização'", () => {
        const campusEvents = [{location: undefined, data: undefined, campus: undefined}]
        
        const wrapper = QuasarTest.shallowMount(DashEventCard, {...PROPS_DATA, campusEvents});
        const cityName = wrapper.find('#DashEventCardCampus')
  
        expect(cityName.text()).toBe("Sem localização");
      })
    })
  })
})