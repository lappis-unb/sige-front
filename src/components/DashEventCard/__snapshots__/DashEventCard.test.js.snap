// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashEventCard Snapshot cardTypeStyle Teste da quando prop de estilo é card-active 1`] = `
<div
  class="base"
  id="DashEventCard"
>
  <q-card-stub
    class="evt-card-size card-overflow card-active"
    flat="true"
    tag="div"
  >
    <q-card-section-stub
      class="text-center card-title"
      tag="div"
    >
      <span>
        <q-icon-stub
          name="img:ic_ocorrencia_1.svg"
          tag="i"
        />
      </span>
       
      <strong
        class="card-title"
      >
        Queda de Fase Teste
      </strong>
    </q-card-section-stub>
     
    <q-card-section-stub
      class="text-center q-px-xs"
      tag="div"
    >
      <p
        class="q-my-xs"
        id="DashEventCardCampus"
      >
        
        Gama (FGA) - A,B,C
      
      </p>
       
      <!---->
    </q-card-section-stub>
  </q-card-stub>
</div>
`;

exports[`DashEventCard Snapshot cardTypeStyle Teste da quando prop de estilo é card-active-critical 1`] = `
<div
  class="base"
  id="DashEventCard"
>
  <q-card-stub
    class="evt-card-size card-overflow card-active-critical"
    flat="true"
    tag="div"
  >
    <q-card-section-stub
      class="text-center card-title"
      tag="div"
    >
      <span>
        <q-icon-stub
          name="img:ic_ocorrencia_1.svg"
          tag="i"
        />
      </span>
       
      <strong
        class="card-title"
      >
        Queda de Fase Teste
      </strong>
    </q-card-section-stub>
     
    <q-card-section-stub
      class="text-center q-px-xs"
      tag="div"
    >
      <p
        class="q-my-xs"
        id="DashEventCardCampus"
      >
        
        Gama (FGA) - A,B,C
      
      </p>
       
      <!---->
    </q-card-section-stub>
  </q-card-stub>
</div>
`;

exports[`DashEventCard Snapshot cardTypeStyle Teste da quando prop de estilo é card-active-warning 1`] = `
<div
  class="base"
  id="DashEventCard"
>
  <q-card-stub
    class="evt-card-size card-overflow card-active-warning"
    flat="true"
    tag="div"
  >
    <q-card-section-stub
      class="text-center card-title"
      tag="div"
    >
      <span>
        <q-icon-stub
          name="img:ic_ocorrencia_1.svg"
          tag="i"
        />
      </span>
       
      <strong
        class="card-title"
      >
        Queda de Fase Teste
      </strong>
    </q-card-section-stub>
     
    <q-card-section-stub
      class="text-center q-px-xs"
      tag="div"
    >
      <p
        class="q-my-xs"
        id="DashEventCardCampus"
      >
        
        Gama (FGA) - A,B,C
      
      </p>
       
      <!---->
    </q-card-section-stub>
  </q-card-stub>
</div>
`;

exports[`DashEventCard Snapshot cardTypeStyle Teste da quando prop de estilo é card-inactive 1`] = `
<div
  class="base"
  id="DashEventCard"
>
  <q-card-stub
    class="evt-card-size card-overflow card-inactive"
    flat="true"
    tag="div"
  >
    <q-card-section-stub
      class="text-center card-title"
      tag="div"
    >
      <span>
        <q-icon-stub
          name="img:ic_ocorrencia_1.svg"
          tag="i"
        />
      </span>
       
      <strong
        class="card-title"
      >
        Queda de Fase Teste
      </strong>
    </q-card-section-stub>
     
    <q-card-section-stub
      class="text-center q-px-xs"
      tag="div"
    >
      <p
        class="q-my-xs"
        id="DashEventCardCampus"
      >
        
        Gama (FGA) - A,B,C
      
      </p>
       
      <!---->
    </q-card-section-stub>
  </q-card-stub>
</div>
`;
